Custom serializers

rubik opened the pull request:
## Changes

I would like to propose this change, which allows the user to specify custom serializers in alternative to pickle, which is left as the default.

The public API is changed only very slightly: there are new `_serialize` and `_deserialize` arguments to `create_pool` and `Worker`. The rest of the changes essentially reduces to passing these to the pickle functions.

For backward compatibility, I have left the `pickle_jobs`, etc. functions in `arq.jobs` with the same name, even though the name is slightly inappropriate now.

## Rationale

The main reason I am proposing this change is to allow better memory usage. The default serialization method, pickle, is quite space inefficient. Since the job data is essentially JSON, we can attain much better memory usage by switching to serialization methods that are more appropriate.

In my case, by using [MsgPack](https://msgpack.org/) instead of Pickle, I am seeing an improvement of 47%-49% in memory usage which is quite significant. At tens to hundreds of thousands of tasks this means hundreds of MBs if not GBs in savings.

## Usage

The usage is quite simple. Existing code works as usual, but one can use a different serialization method as follows:

```python
import msgpack
from arq.connections import create_pool, RedisSettings

pool = await create_pool(RedisSettings(), _serialize=msgpack.packb, _deserialize=msgpack.unpackb)

WorkerSettings:
    serialize = msgpack.packb
    deserialize = msgpack.unpackb
```

## Tests and documentation
The tests pass, except two tests that also fail on master. If this change is approved, I can write a paragraph in the documentation on custom serialization functions.

samuelcolvin commented on arq/connections.py:
```diff
@@ -43,8 +44,24 @@ def __repr__(self):
+    def __init__(
+        self,
+        pool_or_conn,
+        _serialize: Optional[Callable[[Any], bytes]] = None,
```
```suggestion
        _job_serializer: Optional[Callable[[Any], bytes]] = None,
```
Might be even clearer?

samuelcolvin replied:
and same obviously for `_deserialize` -> `_job_deserializer`.

samuelcolvin commented on arq/jobs.py:
```diff
@@ -119,16 +126,30 @@ def __repr__(self):
+    pass
+
+
+def pickle_job(
```
let's rename `unpickle_job` to `deserialize_job`, better to get it right now.

Change should be pretty explicit and easy to fix.

samuelcolvin commented on arq/jobs.py:
```diff
@@ -119,16 +126,30 @@ def __repr__(self):
 
 
-def pickle_job(function_name: str, args: tuple, kwargs: dict, job_try: int, enqueue_time_ms: int):
+class PickleError(SerializationError):
```
remove.

samuelcolvin commented on arq/connections.py:
```diff
@@ -43,8 +44,24 @@ def __repr__(self):
+        self,
+        pool_or_conn,
+        _job_serializer: Optional[Callable[[Any], bytes]] = None,
+        _job_deserializer: Optional[Callable[[bytes], Any]] = None,
```
```suggestion
        _job_deserializer: Optional[Callable[[bytes], Dict[str, Any]] = None,
```
Or am I missing something?

samuelcolvin commented on arq/connections.py:
```diff
@@ -43,8 +44,24 @@ def __repr__(self):
+    def __init__(
+        self,
+        pool_or_conn,
+        _job_serializer: Optional[Callable[[Any], bytes]] = None,
```
```suggestion
        _job_serializer: Optional[Callable[[Dict[str, Any]], bytes]] = None,
```
I think?

samuelcolvin replied:
and obviously elsewhere where it's used, might be worth having these types defined once, eg. in `jobs.py`?

samuelcolvin commented on arq/jobs.py:
```diff
@@ -119,19 +126,29 @@ def __repr__(self):
+    job_try: int,
+    enqueue_time_ms: int,
+    *,
+    _serialize: Optional[Callable[[Any], bytes]] = None,
```
```suggestion
    serialize: Optional[Callable[[Any], bytes]] = None,
```
since this is explicitly required as a keyword arg, I don't think you need to make it "private".

samuelcolvin commented on tests/test_worker.py:
```diff
@@ -449,3 +451,26 @@ class Foo:
+async def test_incompatible_serializers_1(arq_redis_msgpack: ArqRedis, worker):
+    await arq_redis_msgpack.enqueue_job('foobar', _job_id='job_id')
+    worker: Worker = worker(functions=[foobar])
+    with pytest.raises(SerializationError) as exc_info:
```
what's the error here? Might be good to add `assert str(exc_info.value) == ...` just to document what the error is.

And below.

rubik commented:
@samuelcolvin I addressed your suggestions. You said that `serializer` should be public and therefore starting without underscore for the functions in `arq.jobs`. Following that reasoning, the same correction needs to be applied to `arq.connections.create_pool`. That also makes the arguments equal to the parameters in `arq.worker.Worker`. That was done in a separate commit, https://github.com/samuelcolvin/arq/pull/143/commits/0cdaf4bbb0893b5d8c769d7e248cb4e9a4ee2f4a, so if you don't agree I can revert it.

I don't think the HISTORY.rst change belongs to this PR and you should do it in a separate commit, also because I don't know when you will release the next version (even though I hope very soon).

samuelcolvin commented:
The way I work every commit to code has an entry in `HISTORY.rst`, it's the case for all projects so I'm afraid it's a requirement.

Please add the following to history

```
v0.17 (unreleased)
..................
* custom serializers, eg. to use msgpack rather than pickle, #143 by @rubik 

```

samuelcolvin commented on tests/test_worker.py:
```diff
@@ -449,3 +451,32 @@ class Foo:
+    worker: Worker = worker(functions=[foobar])
+    with pytest.raises(SerializationError) as exc_info:
+        await worker.main()
+        assert exc_info.value.startswith('unable to deserialize job: ')
```
```suggestion
    assert exc_info.value.startswith('unable to deserialize job: ')
```
this is indented wrongly I think.

samuelcolvin commented on tests/test_worker.py:
```diff
@@ -449,3 +451,32 @@ class Foo:
+    )
+    with pytest.raises(SerializationError) as exc_info:
+        await worker.main()
+        assert exc_info.value.startswith('unable to deserialize job: ')
```
```suggestion
    assert exc_info.value.startswith('unable to deserialize job: ')
```

rubik commented:
@samuelcolvin Done. Then after you merge this PR I'll add it for the other one.

rubik commented:
@samuelcolvin I fixed the code as per your latest feedback.

rubik commented:
It seems some tests fail on Python 3.8, but I don't think the errors are related to this PR.

samuelcolvin commented on arq/connections.py:
```diff
@@ -43,8 +44,24 @@ def __repr__(self):
     Thin subclass of ``aioredis.Redis`` which adds :func:`arq.connections.enqueue_job`.
+
+    :param redis_settings: an instance of ``arq.connections.RedisSettings``.
+    :param _job_serializer: a function that serializes Python objects to bytes, defaults to pickle.dumps
```
```suggestion
    :param job_serializer: a function that serializes Python objects to bytes, defaults to pickle.dumps
```

samuelcolvin commented on arq/connections.py:
```diff
@@ -43,8 +44,24 @@ def __repr__(self):
+
+    :param redis_settings: an instance of ``arq.connections.RedisSettings``.
+    :param _job_serializer: a function that serializes Python objects to bytes, defaults to pickle.dumps
+    :param _job_deserializer: a function that deserializes bytes into Python objects, defaults to pickle.loads
```
```suggestion
    :param job_deserializer: a function that deserializes bytes into Python objects, defaults to pickle.loads
```

rubik replied:
@samuelcolvin My bad, it's fixed.

samuelcolvin commented:
Great, thanks. I'll deploy as soon as the other PR is passing and I can merge it.

rubik commented:
@samuelcolvin Thanks for taking the time to review it multiple times.
